### PR TITLE
feat: Tally Migration

### PR DESCRIPTION
**Pull request by adityahase**
Screenshot 2019-04-29 at 5 59 57 AM

<details>
<summary>New Tally Migration</summary>
<p>

Screenshot 2019-04-29 at 5 37 42 AM

</p>
</details>

<details>
<summary>Upload master data file (Ready to process master data file)</summary>
<p>

To obtain Master Data File 

1. Go to Gateway of Tally > Display > List of Accounts
1. In the List of Accounts screen press E: Export.
1. Select Format : XML (Data Interchange)
1. Select Type of Masters: All Masters
1. Compress XML file using a zip utility

Screenshot 2019-04-29 at 5 37 52 AM

</p>
</details>

<details>
<summary>Processing master data file</summary>
<p>

Screenshot 2019-04-29 at 5 38 13 AM

</p>
</details>

<details>
<summary>Ready to import master data</summary>
<p>

Name of the company will be obtained from the Master Data file.
Screenshot 2019-04-29 at 5 38 34 AM

</p>
</details>

<details>
<summary>Importing master data</summary>
<p>

Screenshot 2019-04-29 at 5 40 32 AM

</p>
</details>

<details>
<summary>Set Default Warehouse, Default Cost Center and Round Off Account</summary>
<p>

Screenshot 2019-04-29 at 5 41 36 AM

</p>
</details>

<details>
<summary>Upload Day Book data file (Ready to process day book data file)</summary>
<p>

Screenshot 2019-04-29 at 5 41 51 AM

</p>
</details>

<details>
<summary>Processing day book data</summary>
<p>

Screenshot 2019-04-29 at 5 42 50 AM

</p>
</details>

<details>
<summary>Ready to import day book data</summary>
<p>

Screenshot 2019-04-29 at 5 43 02 AM

</p>
</details>

<details>
<summary>Importing day book data</summary>
<p>

Screenshot 2019-04-29 at 5 45 51 AM

</p>
</details>

### CODE REVIEW

**Comment by test2a** [comment]
This is great. Finally big businesses can move to erpnext without having to start everything from scratch

File: erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+					party_type = "Supplier"
+					parties.append({
+						"doctype": party_type,
+						"supplier_name": account.NAME.string,
+						"pan": account.INCOMETAXNUMBER.string if account.INCOMETAXNUMBER else None,
+						"supplier_group": "All Supplier Groups",
+						"supplier_type": "Individual",
+					})
+				if party_type:
+					address = "\n".join([a.string for a in account.find_all("ADDRESS")[:2]])
```
Comment (karthikeyan5) [inline erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py]: Taking only the first two lines of address can lead to losing information... tally allows unlimited lines for the address. can you please take the whole address and put it into `address_line1` and `address_line2`.

**Review by karthikeyan5 — APPROVED** [review]

**Comment by Swad** [comment]
Can you plz add steps to export daybook data from Tally?

**Comment by shreerangamvinay** [comment]
I am getting this error while importing data from tally 


{'event': None, 'retry': 0, 'log': <function log at 0x7faa5d72d050>, 'site': u'testinterio.profezzo.com', 'job_name': u'frappe.utils.background_jobs.run_doc_method', 'method_name': u'frappe.utils.background_jobs.run_doc_method', 'method': <function run_doc_method at 0x7faa5e35f7d0>, 'user': u'interionation@yahoo.com', 'kwargs': {'doc_method': u'_process_master_data', 'name': u'Tally Migration on 16-10-2019 21:57:13', 'doctype': u'Tally Migration'}, 'is_async': True}
Traceback (most recent call last):
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/background_jobs.py", line 99, in execute_job
    method(**kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/background_jobs.py", line 78, in run_doc_method
    getattr(frappe.get_doc(doctype, name), doc_method)(**kwargs)
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py", line 214, in _process_master_data
    items, uoms = get_stock_items_uoms(collection)
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py", line 194, in get_stock_items_uoms
    "stock_uom": item.BASEUNITS.string,
AttributeError: 'NoneType' object has no attribute 'string'

**Comment by anshulbhardwaj5** [comment]
Title
Tally Migration Error
Error
Data:
"Processing Master Data"
----------------------------------------------------------------------------------------------------

Exception:
Traceback (most recent call last):
  File "/opt/erpnext/erpnext/apps/erpnext/erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py", line 249, in _process_master_data
    collection = self.get_collection(self.master_data)
  File "/opt/erpnext/erpnext/apps/erpnext/erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py", line 72, in get_collection
    master = bs(sanitize(emptify(content)), "xml")
UnboundLocalError: local variable 'content' referenced before assignment

**Comment by anshulbhardwaj5** [comment]
Can anyone tell me how to migrate the data to ERPNext in .XML format


Title
Tally Migration Error
Error
Data:
"Processing Master Data"
----------------------------------------------------------------------------------------------------

Exception:
Traceback (most recent call last):
  File "/opt/erpnext/erpnext/apps/erpnext/erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py", line 249, in _process_master_data
    collection = self.get_collection(self.master_data)
  File "/opt/erpnext/erpnext/apps/erpnext/erpnext/erpnext_integrations/doctype/tally_migration/tally_migration.py", line 73, in get_collection
    collection = master.BODY.IMPORTDATA.REQUESTDATA
AttributeError: 'NoneType' object has no attribute 'IMPORTDATA'

**Comment by Muzzy73** [comment]
Hi @adityahase  Can you add it to hooks.py as it is not viewable in Integration. It can only be accessed through awesome bar. Would nice for user to be able to fine it easily. Am using ERPNext: v12.23.0 (version-12), Frappe Framework: v12.20.0 (version-12)

Thank you.

**Pull request merged**